source/MaterialXGenOsl/Nodes/MaterialNodeOsl.cpp

```diff
@@ -0,0 +1,73 @@
+//
+// TM & (c) 2022 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#include <MaterialXGenOsl/Nodes/MaterialNodeOsl.h>
+#include <MaterialXGenShader/ShaderGenerator.h>
+#include <MaterialXGenShader/Shader.h>
+#include <MaterialXGenShader/GenContext.h>
+
+MATERIALX_NAMESPACE_BEGIN
+
+ShaderNodeImplPtr MaterialNodeOsl::create()
+{
+    return std::make_shared<MaterialNodeOsl>();
+}
+
+void MaterialNodeOsl::emitFunctionDefinition(const ShaderNode&, GenContext& context, ShaderStage& stage) const
+{
+    const static string FUNCTION_DEFINITION =
+        "MATERIAL mx_surfacematerial(surfaceshader surface, displacementshader disp)\n"
+        "{\n"
+        "    float opacity_weight = clamp(surface.opacity, 0.0, 1.0);\n"
+        "    return (surface.bsdf + surface.edf) * opacity_weight + transparent() * (1.0 - opacity_weight);\n"
+        "}\n\n";
+
+    const ShaderGenerator& shadergen = context.getShaderGenerator();
+    shadergen.emitBlock(FUNCTION_DEFINITION, FilePath(), context, stage);
```

Comment:
Material node is now responsible for composing the `surfaceshader` components into a single closure.